internal/provider/newrelic/provider_integration_test.go

```diff
@@ -123,12 +132,153 @@ func Test_Getting_external_metric_generates_a_query_not_rejected_by_backend(t *t
 	})
 }
 
-func newrelicProviderWithMetric(t *testing.T, metric newrelic.Metric) provider.ExternalMetricsProvider {
-	t.Helper()
+func Test_Getting_backend_errors_ends_in_metrics_adapter_and_k8s_errors(t *testing.T) {
+	t.Parallel()
 
-	testEnv := testutil.TestEnv{}
+	ctx := testutil.ContextWithDeadline(t)
+
+	testEnv := testutil.TestEnv{
+		StartKubernetes: true,
+		Region:          "Local",
+	}
 	testEnv.Generate(t)
 
+	p := newrelicProviderWithMetric(t, newrelic.Metric{
+		Query: testIntegrationQuery,
+	}, &testEnv)
+
+	runHTTPServer(t, 3000, func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(http.StatusExpectationFailed)
+	})
+
+	runAdapter(t, &testEnv, p)
+
+	m := provider.ExternalMetricInfo{Metric: testMetricName}
+
+	if _, err := p.GetExternalMetric(ctx, "", nil, m); err == nil {
+		t.Fatalf("Error expected")
+	}
+
+	// url reflecting the external metrics endpoint for the test_metrics so that later we can simulate the request a hpa
+	// would perform to fetch metrics.
+	url := fmt.Sprintf("%s/apis/external.metrics.k8s.io/v1beta1/namespaces/*/test_metric", testEnv.BaseURL)
+
+	checkStatusCodeNotOK(ctx, t, &testEnv, url)
+}
+
+// checkStatusCodeNotOK sends a GET request to the given URL.
+//
+// If 401 or 403 return code is received, function will retry.
+//
+// If a status code = 2xx is received, function will fail the given test.
+func checkStatusCodeNotOK(ctx context.Context, t *testing.T, testEnv *testutil.TestEnv, url string) {
+	t.Helper()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		t.Fatalf("Creating request: %v", err)
+	}
+
+	retryUntilFinished(func() bool {
+		resp, err := testEnv.HTTPClient.Do(req)
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+				t.Fatalf("Test timed out: %v", err)
+			}
+
+			t.Logf("Fetching %s: %v", url, err)
+
+			time.Sleep(1 * time.Second)
+
+			return false
+		}
+
+		defer func() {
+			_ = resp.Body.Close()
+		}()
+
+		switch resp.StatusCode {
+		// Generic API server does not wait for RequestHeaderAuthRequestController informers cache to be synchronized,
+		// so until this is done, metrics adapter will be responding with HTTP 403, so we want to retry on that.
+		case http.StatusForbidden, http.StatusUnauthorized:
+			t.Logf("Got %d response code, expected %d: %v. Retrying.", resp.StatusCode, http.StatusOK, resp)
+
+			return false
+		default:
+		}
+
+		return true
+	})
+
+	resp, err := testEnv.HTTPClient.Do(req)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	if isResponseSuccess(resp) {
+		t.Fatalf("Got %d response code, expected != 2xx", resp.StatusCode)
+	}
+}
+
+func retryUntilFinished(f func() bool) {
+	for {
+		if f() {
+			break
+		}
+	}
+}
```

Comment:
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait might be worth checking, as we are already using it in some places in this codebase.